### PR TITLE
determine the key string from either key id or key in the basetrackcache

### PR DESCRIPTION
**Pull request by m0dB**
Fixes https://github.com/mixxxdj/mixxx/issues/13686

Under certain circumstances, when sorting tracks, `BaseTrackCache::findSortInsertionPoint` would be called and in turn call `BaseTrackCache::compareColumnValues`. When comparing the Key column, it would compare the string value (KEY column) which could be empty, instead of taking the KEY_ID column into account. As a result, findSortInsertionPoint would return a wrong value, resulting in https://github.com/mixxxdj/mixxx/issues/13686 

This fix uses the code that was originally in `BaseTrackTableModel::roleValue`, now moved to KeyUtils, to determine the Key string value based on both columns.

Note that this is not the most efficient: we potentially go from numeric key code to string, only to go back to numeric key code in `BaseTrackCache::compareColumnValues`. But I found that avoiding that would have been much more invasive.

### CODE REVIEW

**Comment by fwcd** [comment]
Seems to work for me, though I still can't reproduce the bug reliably using your steps on main/2.5 builds. I'm afraid that I don't have the time to dig into this in detail, so ideally someone more familiar with the key sorting algorithm should review this

File: src/library/basetrackcache.cpp [hunk 1]
```diff
@@ -381,6 +381,8 @@ void BaseTrackCache::getTrackValueForColumn(TrackPointer pTrack,
         trackValue.setValue(pTrack->getRating());
     } else if (fieldIndex(ColumnCache::COLUMN_LIBRARYTABLE_KEY) == column) {
         trackValue.setValue(pTrack->getKeyText());
+        QVariant keyCodeValue(static_cast<int>(pTrack->getKey()));
+        trackValue = KeyUtils::keyFromColumns(std::move(trackValue), std::move(keyCodeValue));
```
Comment (ywwg) [inline src/library/basetrackcache.cpp:385]: I do not think this use of std::move and rvalue reference in the called function do what you think it will do, and it adds noise to the code.  Almost certainly, given the function signatures, the compiler will do the right thing and avoid extra copies by itself.  The function should just take const& and it will work fine
```suggestion
        trackValue = KeyUtils::keyFromColumns(trackValue, keyCodeValue);
```

File: src/track/keyutils.cpp [hunk 2]
```diff
@@ -763,3 +767,34 @@ int KeyUtils::keyToCircleOfFifthsOrder(mixxx::track::io::key::ChromaticKey key,
...
+    // to determine the Key string from either the LIBRARYTABLE_KEY or the
+    // LIBRARYTABLE_KEY_ID field.
+    //
+    // If we know the semantic key via the LIBRARYTABLE_KEY_ID
+    // column (as opposed to the string representation of the key
+    // currently stored in the DB) then lookup the key and render it
+    // using the user's selected notation.
+    if (keyCodeValue.isNull()) {
+        // Otherwise, just use the column value as is
+        return std::move(rawValue);
```
Comment (ywwg) [inline src/track/keyutils.cpp:783]:
```suggestion
QVariant KeyUtils::keyFromColumns(const QVariant& rawValue, const QVariant& keyCodeValue) {
    // Helper function used by basetrackcache.cpp and basetracktablemodel.cpp
    // to determine the Key string from either the LIBRARYTABLE_KEY or the
    // LIBRARYTABLE_KEY_ID field.
    //
    // If we know the semantic key via the LIBRARYTABLE_KEY_ID
    // column (as opposed to the string representation of the key
    // currently stored in the DB) then lookup the key and render it
    // using the user's selected notation.
    if (keyCodeValue.isNull()) {
        // Otherwise, just use the column value as is
        return rawValue;
```

File: src/track/keyutils.h [hunk 3]
```diff
@@ -131,6 +132,8 @@ class KeyUtils {
     static int keyToCircleOfFifthsOrder(mixxx::track::io::key::ChromaticKey key,
                                         KeyNotation notation);
 
+    static QVariant keyFromColumns(QVariant&& rawValue, QVariant&& keyCodeValue);
```
Comment (ywwg) [inline src/track/keyutils.h:135]:
```suggestion
    static QVariant keyFromColumns(const QVariant& rawValue, const QVariant& keyCodeValue);
```

**Review by ywwg — CHANGES_REQUESTED** [review]
https://devblogs.microsoft.com/oldnewthing/20231124-00/?p=109059

File: src/library/basetrackcache.cpp (hunk 1 above)
Comment (m0dB) [reply src/library/basetrackcache.cpp:385]: Yeah, I got carried a way a bit here 😄

**Comment by m0dB** [comment]
Thanks for the suggestions. @ywwg . I applied them and did some additional cleanup. Please re-review.

Thanks for testing @fwcd . Yes, reproducing the bug, even with the steps provided, is not very clear, and neither is the code path that eventually leads to the code with the bug. The whole library code is really hard to follow and seems needlessly complex.

**Review by ywwg — COMMENTED** [review]
is there a unit test we can write for this?

**Comment by m0dB** [comment]
> is there a unit test we can write for this?

If you mean a test for the sorting of the library by key: I am not very familiar with what tests we have, but I get the impression we don't have any tests that populate a testing mixxxdb.sqlite. So I would say that is outside the scope of this bug fix.

If you mean a unit test for the KeyUtils::keyFromColumns function that I isolated from BaseTrackTableModel::roleValue, sure, that could be done easily.

**Comment by daschuer** [comment]
I can confirm the bug. I have also found a couple of tracks in my library where the key text is an empty string (not NULL) but a valid key ID. I have other tracks where the key column contain not recognized strings but a "0" key ID. 

I have no idea how this happens, 

I have checked with the unrecognized track and mixxx 2.6-alpha and the it is imported with an empty string and key ID 0 which is reasonable and does not lead to the bug. 

We need to keep an eye of it id the root cause for the library inconstancy is still there. This reminds me by the we to this pending PR I need to finish. https://github.com/mixxxdj/mixxx/pull/11098

**Review by daschuer — CHANGES_REQUESTED** [review]

File: src/track/keyutils.cpp [hunk 4]
```diff
@@ -763,3 +767,34 @@ int KeyUtils::keyToCircleOfFifthsOrder(mixxx::track::io::key::ChromaticKey key,
...
+    // to determine the Key string from either the LIBRARYTABLE_KEY or the
+    // LIBRARYTABLE_KEY_ID field.
+    //
+    // If we know the semantic key via the LIBRARYTABLE_KEY_ID
+    // column (as opposed to the string representation of the key
+    // currently stored in the DB) then lookup the key and render it
+    // using the user's selected notation.
+    if (keyCodeValue.isNull()) {
+        // Otherwise, just use the column value as is
+        return rawValue;
```
[the remark this answers is not in the document]
Comment (m0dB) [reply src/track/keyutils.cpp:783]: > Since this is only the exceptional path (right?)

Correct.

File: src/track/keyutils.cpp [hunk 5]
```diff
@@ -763,3 +767,34 @@ int KeyUtils::keyToCircleOfFifthsOrder(mixxx::track::io::key::ChromaticKey key,
         return s_sortKeysCircleOfFifthsLancelot[static_cast<int>(key)];
     }
 }
+
+// static
+QVariant KeyUtils::keyFromColumns(const QVariant& rawValue, const QVariant& keyCodeValue) {
```
[the remark this answers is not in the document]
Comment (m0dB) [reply src/track/keyutils.cpp:772]: renamed

**Comment by m0dB** [comment]
done, please re-review and resolve.

**Review by daschuer — APPROVED** [review]
LGTM, Thank you.

**Comment by daschuer** [comment]
@ywwg It looks like your change request is also included. Merge?

**Review by ywwg — APPROVED** [review]
